netket/driver/qsr.py

```diff
@@ -0,0 +1,384 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from functools import partial
+
+import numpy as np
+import jax
+import jax.numpy as jnp
+
+from numba import njit
+
+from netket import jax as nkjax
+from netket.operator import AbstractOperator, LocalOperator
+from netket.stats import Stats
+from netket.vqs import MCState
+from netket.optimizer import (
+    identity_preconditioner,
+    PreconditionerT,
+)
+from netket.utils import mpi
+from netket.utils.types import PyTree
+
+from .vmc_common import info
+from .abstract_variational_driver import AbstractVariationalDriver
+
+
+def _check_bases_type(Us):
+    """Checks if the given bases are valid for the Qsr driver.
+
+    Args:
+        Us (list or np.ndarray): A list of bases
+
+    Raises:
+        ValueError: When not given a list or np.ndarray
+        TypeError: If the type of the operators is not a child of AbstractOperator
+    """
+    if not (isinstance(Us, list) or isinstance(Us, np.ndarray)):
+        raise ValueError(
+            "The bases should be a list or np.ndarray(dtype=object)" " of the bases."
+        )
+
+    if isinstance(Us[0], AbstractOperator):
+        return Us
+
+    raise TypeError("Unknown type of measurement basis.")
+
+
+def _convert_data(sigma_s, Us):
+    """Converts samples and rotation operators to a more direct computational format.
```

Comment:
can you please put a comment on what is this conversion doing/ why is it needed?

Reply:
As far as I understand, it is necessary to achieve a good performance since it is a representation that is handy for the use of jax.mask. However I think it may be not necessary anymore/reviewed anyway when we switch to jax.sparse. Maybe @PhilipVinc can comment?

Reply:
The idea is that each measurement is a tuple of an operator and a bit string.
So it's more efficient to call `get_conn` on every operator with it's bit string once, during setup, instead of at every iteration of QSR.
This simply builds a huge `mels` `sections` `sigma` for all measurements, which then is used to generate minibatches of measurements during training

Reply:
Maybe the name could be improved. Something like `_preprocess_measurements` or `_preprocess_data`.
Whatever the storage format, we'll need somet preprocessing to convert the user-provided measurements to some more efficient format...